lib/membrane_protocol_icecast/input/controller.ex

```diff
@@ -0,0 +1,66 @@
+defmodule Membrane.Protocol.Icecast.Input.Controller do
+  alias Membrane.Protocol.Icecast.Types
+
+  @type state_t :: any
+
+  @type payload_t :: binary
+  @type invalid_reason_t ::
+    {:request, {:request, any} | {:header, binary}} |
+    {:method, atom | charlist} |
+    {:mount, binary} |
+    :format_unknown |
+    :format_not_allowed |
+    :request |
+    :too_many_headers |
+    :unauthorized
+    # FIXME method -> method_*
+
+
+  @type incoming_reply_t :: {:ok, {:allow, state_t} | {:deny, :forbidden}}
+  @type source_reply_t :: {:ok, {:allow, state_t} | {:deny, :unauthorized | :forbidden}}
+
+  @type payload_reply_t :: {:ok, {:continue, state_t} | :drop}
+  @type metadata_reply_t :: {:ok, {:continue, state_t} | :drop}
+
+  # Initial actions
+  @callback handle_init(any) :: {:ok, state_t}
```

Comment:
Specify type for options, add behaviour for transport

Reply:
This is an argument to the `controller_module` which can be of `any` type so we cannot narrow it down any further. I agree with a need for transport behaviour/type. We can put the type in spec to `init/1` function in machine modules.